# IC 4.9.1 Adding Customer Data Set

### 0. List of Contents
#### 1. Importing Libraries and Dataframes 
#### 2. Data Wrangling on Customers
#### 3. Quality and Consistency Checks on Customers
#### 4. Comibining Customers with the other DF
#### 5. Exporting new DF

### 1. Importing Libraries and Dataframes

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Creating path
path = r'/Users/fruzsinasari/Documents/Fru Python/Instacart Basket Analysis/02 Data/'

In [4]:
# Importing customer df
customers = pd.read_csv(os.path.join(path, 'Original Data', 'customers.csv'))

### 2. Data Wrangling on Customers

#### Exploring Customers Data

In [5]:
customers.head(20)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514

In [6]:
customers.tail()

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

In [7]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [8]:
customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

#### The name of the users aren't relevant for my analysis, hence I will drop that

In [9]:
customers=customers.drop(columns=['First Name','Surnam'])

In [10]:
#checking result
customers.head()

user_id  Gender       STATE  Age date_joined  n_dependants fam_status  \
0    26711  Female    Missouri   48    1/1/2017             3    married   
1    33890  Female  New Mexico   36    1/1/2017             0     single   
2    65803    Male       Idaho   35    1/1/2017             2    married   
3   125935  Female        Iowa   40    1/1/2017             0     single   
4   130797  Female    Maryland   26    1/1/2017             1    married   

   income  
0  165665  
1   59285  
2   99568  
3   42049  
4   40374

#### Changing the name of n_dependants column because it is confusing

In [11]:
customers.rename(columns = {'n_dependants' : 'add_ppl_in_household'}, inplace = True)

In [12]:
#checking result
customers.head()

user_id  Gender       STATE  Age date_joined  add_ppl_in_household  \
0    26711  Female    Missouri   48    1/1/2017                     3   
1    33890  Female  New Mexico   36    1/1/2017                     0   
2    65803    Male       Idaho   35    1/1/2017                     2   
3   125935  Female        Iowa   40    1/1/2017                     0   
4   130797  Female    Maryland   26    1/1/2017                     1   

  fam_status  income  
0    married  165665  
1     single   59285  
2    married   99568  
3     single   42049  
4    married   40374

#### In this df user_id is int64, this might have to be changed to object depending on the other df -> will check after other df is imported

### 3. Quality and Consistency Checks on Customers

#### Missing values

In [13]:
customers.isnull().sum()

user_id                 0
Gender                  0
STATE                   0
Age                     0
date_joined             0
add_ppl_in_household    0
fam_status              0
income                  0
dtype: int64

#### There are no missing values, no action needs to be taken. Besides based on the earlier run .describe() syntax all values make sense there too.

#### Duplicates

In [14]:
customers_dups=customers[customers.duplicated()]

In [15]:
customers_dups

Empty DataFrame
Columns: [user_id, Gender, STATE, Age, date_joined, add_ppl_in_household, fam_status, income]
Index: []

#### There are no full duplicates and since we have 206209 entries and the min of user_id is 1 and the max is 206209 then probably we do not have duplicate user_ids either.

#### Mixed-Type Data

In [16]:
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

#### There are no mixed-type data in the df

### 4. Comibining Customers with the other DF

#### Importing the other df

In [17]:
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Prepared Data', 'orders_products_with_flags_v2.pkl'))

In [18]:
# comlpeting the missing cleaning step from the exercise, so that ords_prods_merge is clean and accurate
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [19]:
# checking result 
ords_prods_merge['prices'].max()

25.0

In [21]:
# checking ords_prods_merge df and datatypes
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32435059 entries, 0 to 32435058
Data columns (total 23 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   orders_day_of_week      int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   product_id              int64   
 7   add_to_cart_order       int64   
 8   reordered               int64   
 9   product_name            object  
 10  aisle_id                float64 
 11  department_id           float64 
 12  prices                  float64 
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  Busiest_period_of_day   object  
 17  max_order               int64   
 18  loyalty_flag            object  
 19  avg_price_per_user      float64 
 20  spending_flag           object  
 21  median

#### user_id, which is our key is int64 in both df

In [23]:
ords_prods_merge['user_id'].describe()

count    3.243506e+07
mean     1.029372e+05
std      5.946651e+04
min      1.000000e+00
25%      5.142000e+04
50%      1.026110e+05
75%      1.543910e+05
max      2.062090e+05
Name: user_id, dtype: float64

#### ords_prods_merge has the same max for user_id as customers this means that user_id is a fully matching column

In [30]:
#removing _merge column in preparation for merge
ords_prods_merge=ords_prods_merge.drop(columns=['_merge'])

In [31]:
ords_prods_cust_merge = pd.merge(ords_prods_merge, customers, on = 'user_id', indicator = True)

In [32]:
#Confirm the results of the merge using the merge flag
ords_prods_cust_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  ...  median_prior_order  \
0                                     Soda  ...                20.5   
1  Organic Unsweetened Vanilla Almond Milk  ...                20.5   
2                      Original Beef Jerky  ...                20.5   
3               Aged White Cheddar Popcorn  ...                20.5   
4         XL Pick-A-Size Paper Towel Rolls  ...                20.5   

   frequent_customer_flag  Gender    STATE Age date_joined  \
0   Non-frequent customer  Female  Alabama  31   2/17/2019   
1   Non-frequent customer  Female  Alabama  31   2/17/2019   
2   Non-frequent customer  Female  Alabama  31   2/17/2019   
3   Non-frequent customer  Female  Alabama  31   2/17/2019   
4   Non-frequent customer  Female  Alabama  31   2/17/2019   

   add_ppl_in_household fam_status  income _merge  
0                     3    married   40423   both  
1                     3    married   40423   both  
2                     3    married   40423   both  
3                     3    married   40423   both  
4                     3    married   40423   both  

[5 rows x 30 columns]

In [33]:
#Confirm the results of the merge using the merge flag with frequency
ords_prods_cust_merge['_merge'].value_counts()

both          32435059
left_only            0
right_only           0
Name: _merge, dtype: int64

#### Merge worked based on head syntax and the merge flag, because the ords_prods_merge originally had  32435059 rows.

In [34]:
# _merge flag isn't needed anymore
ords_prods_cust_merge=ords_prods_cust_merge.drop(columns=['_merge'])

### 5. Exporting new DF

In [35]:
# Export data to pkl due to size
ords_prods_cust_merge.to_pickle(os.path.join(path,'Prepared Data', 'ords_prods_cust_merge.pkl'))